src/models/state/wallet/wallet_state.rs

```diff
@@ -1059,7 +1196,46 @@ impl WalletState {
         let spent_inputs: Vec<(Utxo, AbsoluteIndexSet, u64)> =
             self.scan_for_spent_utxos(&tx_kernel).await;
 
-        let onchain_received_outputs = self.scan_for_announced_utxos(&tx_kernel);
+        let onchain_received_outputs = self
+            .scan_for_utxos_announced_to_known_keys(&tx_kernel)
+            .collect_vec(); // drop immutable borrow of self, for (maybe) bumping below
+
+        // if scan mode is active, scan the block with keys that will be
+        // derived in the future
+        let mut outputs_recovered_through_scan_mode = vec![];
+        if let Some(scan_mode_configuration) = self.scan_mode {
+            if scan_mode_configuration.block_is_in_range(new_block) {
+                let mut max_counters = HashMap::<KeyType, u64>::new();
+                for (key_type, derivation_index, incoming_utxo) in self
+                    .scan_for_utxos_announced_to_future_keys(
+                        scan_mode_configuration.num_future_keys(),
+                        &tx_kernel,
+                    )
+                    .await
+                {
+                    if max_counters
+                        .get(&key_type)
+                        .is_none_or(|&current_max_derivation_index| {
+                            current_max_derivation_index < derivation_index
+                        })
+                    {
+                        max_counters.insert(key_type, derivation_index);
+                    }
+                    outputs_recovered_through_scan_mode.push(incoming_utxo);
+                }
+
+                info!(
+                    "Scan Mode: recovered {} UTXOs in block {}",
+                    outputs_recovered_through_scan_mode.len(),
+                    new_block.header().height
+                );
+
+                for (key_type, derivation_index) in max_counters.into_iter() {
+                    self.bump_derivation_counter(key_type, derivation_index)
+                        .await;
+                }
+            }
+        }
```

Comment:
Could this be factored out? Would allow for some nice testing, I think.

Reply:
On second thought: Don't. Since it mutates state, it's not very nice to factor it out, I think.

Reply:
That's why I left this task undone yesterday. I couldn't figure out how to factor it out nicely.

One could factor out a function that wraps `scan_for_utxos_announced_to_future_keys` along with max-tracking, and then test the max-tracking in isolation. That would be an improvement in terms of test coverage, but in terms of code quality maybe not so much.